stdlib/src/bisect.rs

```diff
@@ -66,20 +68,35 @@ mod _bisect {
     #[inline]
     #[pyfunction]
     fn bisect_left(
-        BisectArgs { a, x, lo, hi }: BisectArgs,
+        BisectArgs { a, x, lo, hi, key }: BisectArgs,
         vm: &VirtualMachine,
     ) -> PyResult<usize> {
         let (mut lo, mut hi) = as_usize(lo, hi, a.length(vm)?, vm)?;
 
-        while lo < hi {
-            // Handles issue 13496.
-            let mid = (lo + hi) / 2;
-            if a.get_item(mid, vm)?.rich_compare_bool(&x, Lt, vm)? {
-                lo = mid + 1;
-            } else {
-                hi = mid;
+        if let Some(ref key) = key {
+            while lo < hi {
+                let mid = (lo + hi) / 2;
+                let a_mid = a.get_item(mid, vm)?;
+                let comp = vm.invoke(key, (a_mid.clone(),))?;
+                if comp.rich_compare_bool(&x, Lt, vm)? {
+                    lo = mid + 1;
+                } else {
+                    hi = mid;
+                }
+            }
+
+        } else {
+            while lo < hi {
+                // Handles issue 13496.
+                let mid = (lo + hi) / 2;
+                if a.get_item(mid, vm)?.rich_compare_bool(&x, Lt, vm)? {
+                    lo = mid + 1;
+                } else {
+                    hi = mid;
+                }
```

Comment:
It seems `bisect_left` and `bisect_right` implementation share some duplicated codes except the order of codes `lo = mid = 1;` and `hi = mid;`. How do you think?

Reply:
I think that is regarded as repeated by coincidence. we named everything binary search, but they have a few variants. we don't have much gain by combining them.

Reply:
We don't have much gain by combining them.

Reply:
Thank you for your review!
In my opinion, the duplication is a coincidence.
I think solving value `i` where all `e` in `a[:i]` have `e < x` and solving value `i` where all `e` in `a[:i]` have `e <= x` are different tasks.
Generally speaking, unnecessary dependency may increase maintenance costs.
So, I thought it is better to leave the duplication, but I'm not sure😖